mempool/clist_sidecar.go

```diff
@@ -0,0 +1,655 @@
+			BundleSize:   txInfo.BundleSize,
+			BundleHeight: txInfo.DesiredHeight,
+			BundleOrder:  txInfo.BundleOrder,
+		}
+	}
+
+	// -------- BUNDLE EXISTENCE CHECKS ---------
+
+	var bundle *Bundle
+	// load existing bundle, or MAKE NEW if not
+	existingBundle, _ := sc.bundles.LoadOrStore(Key{txInfo.DesiredHeight, txInfo.BundleID}, &Bundle{
+		DesiredHeight: txInfo.DesiredHeight,
+		BundleID:      txInfo.BundleID,
+		CurrSize:      int64(0),
+		EnforcedSize:  txInfo.BundleSize,
+		// TODO: add from gossip info?
+		GasWanted:     int64(0),
+		OrderedTxsMap: &sync.Map{},
+	})
+	bundle = existingBundle.(*Bundle)
+
+	// -------- BUNDLE SIZE CHECKS ---------
+
+	// check if bundle is asking for a different size than one already stored
+	if txInfo.BundleSize != bundle.EnforcedSize {
+		sc.logger.Info(
+			"failed adding sidecarTx",
+			"reason", "trying to insert a tx for bundle at an order greater than the size of the bundle...",
+			"bundle id", txInfo.BundleID,
+			"bundle size", txInfo.BundleSize,
+			"gasWanted", txInfo.GasWanted,
+			"enforced size", bundle.EnforcedSize,
+			"tx", tx.Hash(),
+		)
+		return ErrTxMalformedForBundle{
+			BundleID:     txInfo.BundleID,
+			BundleSize:   txInfo.BundleSize,
+			BundleHeight: txInfo.DesiredHeight,
+			BundleOrder:  txInfo.BundleOrder,
+		}
+	}
+
+	shouldReturn, err := func() (bool, error) {
+		// TODO: make this a lock per bundle?
+		sc.bundleSizeMtx.Lock()
+		defer sc.bundleSizeMtx.Unlock()
+		// Can't add transactions if the bundle is already full
+		// check if the current size of this bundle is greater than the expected size for the bundle, if so skip
+		if bundle.CurrSize >= bundle.EnforcedSize {
+			sc.logger.Info(
+				"failed adding sidecarTx",
+				"reason", "bundle already full for this BundleID...",
+				"bundle id", txInfo.BundleID,
+				"bundle curr size", bundle.CurrSize,
+				"enforced size", bundle.EnforcedSize,
+				"tx", tx.Hash(),
+			)
+			return false, ErrBundleFull{
+				BundleID:     txInfo.BundleID,
+				BundleHeight: txInfo.BundleSize,
+			}
+		}
+
+		// -------- TX INSERTION INTO BUNDLE ---------
+
+		// get the map of order -> scTx
+		orderedTxsMap := bundle.OrderedTxsMap
+
+		// if we already have a tx at this BundleID, bundleOrder, and height, then skip this one!
+		if _, loaded := orderedTxsMap.LoadOrStore(txInfo.BundleOrder, scTx); loaded {
+			// if we had the tx already, then skip
+			sc.logger.Info(
+				"failed adding sidecarTx",
+				"reason", "already have a tx for this BundleID, height, and bundleOrder...",
+				"bundle id", txInfo.BundleID,
+				"bundle height", txInfo.DesiredHeight,
+				"bundle order", txInfo.BundleOrder,
+				"tx", tx.Hash(),
+			)
+			return true, nil
+		}
+		// if we added, then increment bundle size for BundleID
+		atomic.AddInt64(&bundle.CurrSize, int64(1))
+		return false, nil
+	}()
+	if err != nil {
+		return err
+	} else if shouldReturn {
+		return nil
+	}
+
+	// -------- UPDATE MAX BUNDLE ---------
+
+	func() {
```

Comment:
Anonymous functions harm readability in general. I would not abuse them. In this case in particular it would be trivial to get rid of it and just place `maxBundleIDMtx.Unlock()` after line 304. A different case is the anonymous function defined in line 248, which has several returning points.

Reply:
Listed in https://github.com/informalsystems/audit-skip/issues/6

Reply:
Done in https://github.com/skip-mev/mev-tendermint/pull/131